### PR TITLE
1856 nationalization train discard

### PR DESCRIPTION
**Pull request by ventusignis**
Also adds optional rules and some assorted fixes

### CODE REVIEW

File: lib/engine/game/g_1856/step/discard_train.rb [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+          def time_to_discard
+            @game.nationalization_train_discard_trigger
+          end
+
+          def can_discard(entity)
+            entity == @game.national && time_to_discard && !national_discardable_trains.empty?
+          end
+
+          def must_discard(entity)
+            entity == @game.national && time_to_discard && national_discardable_trains.length > 3
```
Comment (tobymao) [inline lib/engine/game/g_1856/step/discard_train.rb:64]: size

File: lib/engine/game/g_1856/step/discard_train.rb [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+          end
+
+          def national_discardable_trains
+            non_permanent_trains = @game.national.trains.select(&:rusts_on)
+            permanent_trains = @game.national.trains.reject(&:rusts_on)
+            non_permanent_trains + (permanent_trains.length > 3 ? permanent_trains : [])
```
Comment (tobymao) [inline lib/engine/game/g_1856/step/discard_train.rb:70]: size

**Review by tobymao — APPROVED** [review]

File: lib/engine/game/g_1856/step/discard_train.rb (hunk 1 above)
Comment (ventusignis) [reply lib/engine/game/g_1856/step/discard_train.rb]: fixed

File: lib/engine/game/g_1856/step/discard_train.rb (hunk 2 above)
Comment (ventusignis) [reply lib/engine/game/g_1856/step/discard_train.rb]: fixed

**Pull request merged**